storage/interceptors/smaap_service_binding_interceptor.go

```diff
@@ -0,0 +1,521 @@
+type ServiceBindingCreateInterceptorProvider struct {
+	OSBClientCreateFunc osbc.CreateFunc
+	Repository          storage.TransactionalRepository
+	TenantKey           string
+	PollingInterval     time.Duration
+}
+
+func (p *ServiceBindingCreateInterceptorProvider) Provide() storage.CreateAroundTxInterceptor {
+	return &ServiceBindingInterceptor{
+		osbClientCreateFunc: p.OSBClientCreateFunc,
+		repository:          p.Repository,
+		tenantKey:           p.TenantKey,
+		pollingInterval:     p.PollingInterval,
+	}
+}
+
+func (c *ServiceBindingCreateInterceptorProvider) Name() string {
+	return ServiceBindingCreateInterceptorProviderName
+}
+
+const ServiceBindingDeleteInterceptorProviderName = "ServiceBindingDeleteInterceptorProvider"
+
+// ServiceBindingDeleteInterceptorProvider provides an interceptor that notifies the actual broker about instance deletion
+type ServiceBindingDeleteInterceptorProvider struct {
+	OSBClientCreateFunc osbc.CreateFunc
+	Repository          storage.TransactionalRepository
+	TenantKey           string
+	PollingInterval     time.Duration
+}
+
+func (p *ServiceBindingDeleteInterceptorProvider) Provide() storage.DeleteAroundTxInterceptor {
+	return &ServiceBindingInterceptor{
+		osbClientCreateFunc: p.OSBClientCreateFunc,
+		repository:          p.Repository,
+		tenantKey:           p.TenantKey,
+		pollingInterval:     p.PollingInterval,
+	}
+}
+
+func (c *ServiceBindingDeleteInterceptorProvider) Name() string {
+	return ServiceBindingDeleteInterceptorProviderName
+}
+
+type ServiceBindingInterceptor struct {
+	osbClientCreateFunc osbc.CreateFunc
+	repository          storage.TransactionalRepository
+	tenantKey           string
+	pollingInterval     time.Duration
+}
+
+func (i *ServiceBindingInterceptor) AroundTxCreate(f storage.InterceptCreateAroundTxFunc) storage.InterceptCreateAroundTxFunc {
+	return func(ctx context.Context, obj types.Object) (types.Object, error) {
+		binding := obj.(*types.ServiceBinding)
+
+		instance, err := getInstanceByID(ctx, binding.ServiceInstanceID, i.repository)
+		if err != nil {
+			return nil, err
+		}
+
+		if instance.PlatformID != types.SMPlatform {
+			log.C(ctx).Debugf("platform is not %s. Skipping interceptor %s", types.SMPlatform, ServiceBindingDeleteInterceptorProviderName)
+			return f(ctx, obj)
+		}
+		operation, found := operations.GetFromContext(ctx)
+		if !found {
+			return nil, fmt.Errorf("operation missing from context")
+		}
+
+		osbClient, broker, service, plan, err := prepare(ctx, i.repository, i.osbClientCreateFunc, instance)
+		if err != nil {
+			return nil, err
+		}
+
+		if isBindable := !i.isPlanBindable(service, plan); isBindable {
+			return nil, &util.HTTPError{
+				ErrorType:   "BrokerError",
+				Description: fmt.Sprintf("plan %s is not bindable", plan.CatalogName),
+				StatusCode:  http.StatusBadGateway,
+			}
+		}
+
+		if isReady := i.isInstanceReady(instance); !isReady {
```

Comment:
From OSB spec:
`The Platform SHOULD NOT allow the creation of new bindings of an unusable Service Instance until the instance has been deleted or repaired by a subsequent update. `